src/Utilities/MDFlatten.jl

```diff
@@ -0,0 +1,99 @@
+"""
+Proved the [`mdflatten`](@ref) function that can "flatten" Markdown objects into
+a string, with formatting etc. stripped.
+
+Note that the tests in `test/mdflatten.jl` should be considered to be the spec
+for the output (number of newlines, indents, formatting, etc.).
+"""
+module MDFlatten
+
+export mdflatten
+
+import Base.Markdown:
+    MD, BlockQuote, Bold, Code, Header, HorizontalRule,
+    Image, Italic, LaTeX, LineBreak, Link, List, Paragraph, Table
+
+"""
+Convert a Markdown object to a `String` of only text (i.e. not formatting info).
+
+It drop most of the extra information (e.g. language of a code block, URLs)
+and formatting (e.g. emphasis, headers). This "flattened" representation can
+then be used as input for search engines.
+"""
+function mdflatten(md)
+    io = IOBuffer()
+    mdflatten(io, md)
+    takebuf_string(io)
+end
+
+mdflatten(io, md) = mdflatten(io, md, md)
+mdflatten(io, md::MD, parent) = mdflatten(io, md.content, md)
+mdflatten(io, vec::Vector, parent) = map(x->mdflatten(io, x, parent), vec)
+function mdflatten(io, vec::Vector, parent::MD)
+    # this special case separates top level blocks with newlines
+    for md in vec
+        mdflatten(io, md, parent)
+        print(io, "\n\n")
+    end
+end
+
+# Block level MD nodes
+mdflatten{N}(io, h::Header{N}, parent) = mdflatten(io, h.text, h)
+mdflatten(io, p::Paragraph, parent) = mdflatten(io, p.content, p)
+mdflatten(io, bq::BlockQuote, parent) = mdflatten(io, bq.content, bq)
+mdflatten(io, ::HorizontalRule, parent) = nothing
+function mdflatten(io, list::List, parent)
+    for (idx,li) in enumerate(list.items)
+        for (jdx,x) in enumerate(li)
+            mdflatten(io, x, list)
+            jdx != length(li) && print(io, '\n')
```

Comment:
Just to clarify -- the idiom basically is to minimize the number of `!`s?

Just asking because, my own go-to idiom would have been `&&` for normal control flow and `||` for errors (after having thought about it for a sec I think that's too strict in any case, although the "something or error()" part still seems appealing).

(Will make the change in any case, consistency is the key.)

As for `println`: I would maybe stick with `print`, with the reason being that I'm rather picky about what gets outputted when, how many newlines etc. Also with the change we'd have `print`s and `println`s mixed, and I think it would make it less readable. And `print("\n\n")` -- should that be `println('\n')` or `println(); println()`?